src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -701,58 +611,91 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VariantMetadata,
     annotateSamples(s => m.getOrElse(s, null), finalType, inserter)
   }
 
-  def annotateSamplesKeyTable(kt: KeyTable, code: String): VariantSampleMatrix[T] = {
-    if (kt.keyFields.length != 1 || kt.keyFields(0).typ != TString)
-      fatal(s"Annotating samples requires a key table with 1 key of type String. " +
-        s"Found ${ kt.nKeys } ${ plural(kt.nKeys, "key") }${
-          if (kt.nKeys == 0) "" else s": [ ${
-            kt.keyFields.map(f => s"${ prettyIdentifier(f.name) }: ${ f.typ }").mkString(", ")
-          } ]"
-        }")
-
-    val ec = EvalContext(Map("sa" -> (0, saSignature), "table" -> (1, kt.valueSignature)))
+  def annotateSamplesTable(kt: KeyTable, vdsKey: java.util.ArrayList[String],
+    root: String, expr: String, product: Boolean): VariantSampleMatrix[T] =
+    annotateSamplesTable(kt, if (vdsKey != null) vdsKey.asScala else null, root, expr, product)
 
-    val (finalType, inserter) =
-      buildInserter(code, saSignature, ec, Annotation.SAMPLE_HEAD)
+  def annotateSamplesTable(kt: KeyTable, vdsKey: Seq[String] = null,
+    root: String = null, expr: String = null, product: Boolean = false): VariantSampleMatrix[T] = {
 
-    val ktMap = kt.keyedRDD()
-      .map { case (k, v) => (k.asInstanceOf[Row].getAs[String](0), v) }
-      .filter(_._1 != null)
-      .collectAsMap()
+    if (root == null && expr == null || root != null && expr != null)
+      fatal("method `annotateSamplesTable' requires one of `root' or 'expr', but not both")
 
-    annotateSamples(s => ktMap.getOrElse(s, null), finalType, inserter)
-  }
+    var (joinSignature, f): (Type, Annotation => Annotation) = kt.valueSignature.size match {
+      case 0 => (TBoolean, _ != null)
+      case 1 => (kt.valueSignature.fields.head.typ, x => if (x != null) x.asInstanceOf[Row].get(0) else null)
+      case _ => (kt.valueSignature, identity[Annotation])
+    }
 
-  def annotateSamplesKeyTable(kt: KeyTable, vdsKey: java.util.ArrayList[String], code: String): VariantSampleMatrix[T] =
-    annotateSamplesKeyTable(kt, vdsKey.asScala, code)
+    if (product) {
+      joinSignature = if (joinSignature == TBoolean) TInt else TArray(joinSignature)
+      val innerF = f
+      f = if (kt.valueSignature.size == 0)
+        _.asInstanceOf[IndexedSeq[_]].length
```

Comment:
No. The distinct join gives null or a random match, but the product join gives empty array or array of all matches